### PR TITLE
Add normalized rewards and default observation and reward.

### PR DESCRIPTION
**Pull request by ChrisCummins**
* If the compiler service crashes during `env.step()`, a default value for reward and observation is returned.
* Add `Norm`-suffixed rewards that are normalized to unoptimized cost.
* A general tidy-up of the observation spaces API. The new `ObservationSpaceSpec.make_derived_space()` and `ObservationView.add_derived_space()` makes it easier to manage python-side observation spaces.

### CODE REVIEW

File: compiler_gym/envs/llvm/service/Cost.h [hunk 1]
```diff
@@ -56,6 +56,6 @@ void setbaselineCosts(const llvm::Module& unoptimizedModule, BaselineCosts* base
 LlvmCostFunction getCostFunction(LlvmRewardSpace space);
 
 // Translate from reward space to a baseline policy.
```
Comment (hughleat) [inline compiler_gym/envs/llvm/service/Cost.h:58]: Okay, I see. You return None when no baseline is needed. Maybe explain in the comment?

File: compiler_gym/envs/llvm/service/RewardSpaces.h [hunk 2]
```diff
@@ -20,13 +20,16 @@ namespace compiler_gym::llvm_service {
 enum class LlvmRewardSpace {
   // Returns the number of IR instructions in the current module.
```
Comment (hughleat) [inline compiler_gym/envs/llvm/service/RewardSpaces.h:21]: Why comment only this line?

File: compiler_gym/envs/compiler_env.py [hunk 3]
```diff
@@ -525,6 +533,12 @@ def step(self, action: int) -> step_t:
         except (ServiceError, ServiceTransportError, TimeoutError) as e:
             self.close()
             info = {"error_details": str(e)}
+            if self.reward_space:
+                reward = self.reward_space.default_value
+                if self.reward_space.default_negates_returns:
```
Comment (hughleat) [inline compiler_gym/envs/compiler_env.py:537]: Feels a bit odd. Why not something like:
`self.reward_space.reward_on_error(self.eipsode_reward, err_type_or_something)`?

File: compiler_gym/envs/llvm/service/ObservationSpaces.cc [hunk 4]
```diff
@@ -38,13 +45,17 @@ std::vector<ObservationSpace> getLlvmObservationSpaceList() {
         ScalarRange featureSize;
         featureSize.mutable_min()->set_value(0);
         std::vector<ScalarRange> featureSizes;
+        featureSizes.reserve(/* autophase feature vector dim: */ 56);
```
Comment (hughleat) [inline compiler_gym/envs/llvm/service/ObservationSpaces.cc:48]: Make 56 a named constant somewhere?

File: compiler_gym/views/observation_space_spec.py [hunk 5]
```diff
@@ -31,66 +40,160 @@ class ObservationSpaceSpec(object):
...
+        def make_seq(scalar_range, dtype, defaults):
+            return Sequence(
+                size_range=scalar_range2tuple(scalar_range, defaults),
+                dtype=dtype,
+                opaque_data_format=proto.opaque_data_format,
+            )
+
+        # Translate from protocol buffer specification to python. There are
+        # three variables to derive: 'space', the gym.Space instance describing
+        # the space. 'cb' is a callback that translates from an Observation
```
Comment (hughleat) [inline compiler_gym/views/observation_space_spec.py:96]: Ah, that's what cb is. Consider changing the name to indicate its use?

File: compiler_gym/views/observation_space_spec.py [hunk 6]
```diff
@@ -31,66 +40,160 @@ class ObservationSpaceSpec(object):
...
+        # message to a python type. and 'to_string' is a callback that
+        # translates from a python type to a string for printing.
+        if proto.opaque_data_format == "json://networkx/MultiDiGraph":
```
Comment (hughleat) [inline compiler_gym/views/observation_space_spec.py:99]: Would it be sensible to define all these constants somewhere and use by name?

File: compiler_gym/views/observation_space_spec.py [hunk 7]
```diff
@@ -31,66 +40,160 @@ class ObservationSpaceSpec(object):
...
+        space: Optional[Space] = None,
+        deterministic: Optional[bool] = None,
+        default_value: Optional[observation_t] = None,
+        platform_dependent: Optional[bool] = None,
+        to_string: Callable[[observation_t], str] = None,
+    ) -> "ObservationSpaceSpec":
+        """Create a derived observation space.
+
+        :param id: The name of the derived observation space.
+        :param cb: A callback function to compute a derived observation
```
Comment (hughleat) [inline compiler_gym/views/observation_space_spec.py:165]: Hmm. Is `cb` the right name here?

File: compiler_gym/views/observation_space_spec.py [hunk 8]
```diff
@@ -31,66 +40,160 @@ class ObservationSpaceSpec(object):
...
+            from the base observation.
+        :param space: The observation space.
```
Comment (hughleat) [inline compiler_gym/views/observation_space_spec.py:167]: "parent_space" or "base_space"?

File: compiler_gym/views/observation_space_spec.py [hunk 9]
```diff
@@ -31,66 +40,160 @@ class ObservationSpaceSpec(object):
...
+        :param deterministic: Whether the observation space is deterministic.
+            If not provided, the value is inherited from the base observation
+            space.
+        :param default_value: The default value for the observation space. If
```
Comment (hughleat) [inline compiler_gym/views/observation_space_spec.py:171]: "value_on_error"?

File: tests/envs/llvm/service_connection_test.py [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+import compiler_gym  # Register environments.
+from compiler_gym.envs import CompilerEnv
+from compiler_gym.envs.llvm.llvm_env import LlvmEnv
+from compiler_gym.service.connection import CompilerGymServiceConnection
+from compiler_gym.util.runfiles_path import runfiles_path
+from tests.test_main import main
+
+pytest_plugins = ["tests.envs.llvm.fixtures"]
+
+SERVICE_BIN = Path(runfiles_path("CompilerGym/compiler_gym/envs/llvm/service/service"))
```
Comment (hughleat) [inline tests/envs/llvm/service_connection_test.py:21]: This gets defined in multiple files. Maybe define once?

**Review by hughleat — COMMENTED** [review]
LGTM.
Few suggestions.

File: compiler_gym/envs/llvm/service/Cost.h (hunk 1 above)
Comment (ChrisCummins) [reply compiler_gym/envs/llvm/service/Cost.h:58]: 👍  good idea, done

File: compiler_gym/envs/llvm/service/RewardSpaces.h (hunk 2 above)
Comment (ChrisCummins) [reply compiler_gym/envs/llvm/service/RewardSpaces.h:21]: done

**Review by ChrisCummins — COMMENTED** [review]
Sorry @hughleat, I shouldn't have merged this without your review. Great suggestions, I've opened #35 to follow up.

Cheers,
Chris

**Pull request merged**